### PR TITLE
feat: search and filter for employees

### PR DESCRIPTION
**Pull request by datphamcode295**
#### What's this PR do?

- [x] Search and filter for employee

#### What are the relevant Git tickets?
https://www.notion.so/dwarves/Add-search-and-filter-for-employees-efe577c9bb944789b5f3c22b8dd5ed5b

#### Screenshots (if appropriate)

// Use [Licecap](http://www.cockos.com/licecap/) to share a screencast gif.
image

### CODE REVIEW

File: pkg/handler/employee/request.go [hunk 1]
```diff
@@ -11,6 +11,10 @@ type GetListEmployeeQuery struct {
 
 	WorkingStatus string `json:"workingStatus" form:"workingStatus"`
 	Preload       bool   `json:"preload" form:"preload,default=true"`
+	PositionID    string `json:"positionID" form:"positionID"`
+	StackID       string `json:"stackID" form:"stackID"`
+	ProjectID     string `json:"projectID" form:"projectID"`
+	SearchValue   string `json:"searchValue" form:"searchValue"`
```
Comment (nguyennh4522) [inline pkg/handler/employee/request.go:17]: it should be `q` or `keyword`

File: pkg/handler/employee/employee.go [hunk 2]
```diff
@@ -63,10 +69,26 @@ func (h *handler) List(c *gin.Context) {
...
 
-	employees, total, err := h.store.Employee.Search(h.repo.DB(), employee.SearchFilter{
+	searchFilter := employee.SearchFilter{
 		WorkingStatus: query.WorkingStatus,
 		Preload:       query.Preload,
-	}, query.Pagination)
+		SearchValue:   query.SearchValue,
+	}
+
+	if query.PositionID != "" {
```
Comment (nguyennh4522) [inline pkg/handler/employee/employee.go:78]: we don't do that here

File: pkg/routes/v1.go [hunk 3]
```diff
@@ -22,7 +22,7 @@ func loadV1Routes(r *gin.Engine, h *handler.Handler, repo store.DBRepo, s *store
 	v1.PUT("/profile", amw.WithAuth, h.Profile.UpdateInfo)
 
 	// employees
-	v1.GET("/employees", amw.WithAuth, pmw.WithPerm("employees.read"), h.Employee.List)
+	v1.GET("/employees", amw.WithAuth, h.Employee.List) //pmw.WithPerm("employees.read"),
```
Comment (nguyennh4522) [inline pkg/routes/v1.go:25]: why?

File: pkg/store/employee/request.go [hunk 4]
```diff
@@ -10,6 +10,10 @@ type SearchFilter struct {
 	EmployeeID    string
 	WorkingStatus string
 	Preload       bool
+	PositionID    model.UUID
```
Comment (nguyennh4522) [inline pkg/store/employee/request.go:13]: should use string

File: pkg/store/employee/employee.go [hunk 5]
```diff
@@ -51,6 +51,23 @@ func (s *store) Search(db *gorm.DB, filter SearchFilter, pagination model.Pagina
...
+		query = query.Joins("JOIN employee_stacks ON employees.id = employee_stacks.employee_id AND employee_stacks.stack_id = ?", filter.StackID)
+	}
+
+	if !filter.ProjectID.IsZero() {
+		query = query.Joins("JOIN project_members ON employees.id = project_members.employee_id AND project_members.project_id = ?", filter.ProjectID)
+	}
+
+	if filter.SearchValue != "" {
+		query = query.Where("github_id like ? OR discord_id like ? OR notion_id like ? OR full_name like ? OR team_email like ?",
+			"%"+filter.SearchValue+"%", "%"+filter.SearchValue+"%", "%"+filter.SearchValue+"%", "%"+filter.SearchValue+"%", "%"+filter.SearchValue+"%")
```
Comment (nguyennh4522) [inline pkg/store/employee/employee.go:68]: use fmt.Sprintf() instead of`"..."+abc+"..."`

File: pkg/routes/v1.go (hunk 3 above)
Comment (datphamcode295) [reply pkg/routes/v1.go]: just for developing purpose, I updated it

File: pkg/store/employee/employee.go [hunk 6]
```diff
@@ -51,6 +53,23 @@ func (s *store) Search(db *gorm.DB, filter SearchFilter, pagination model.Pagina
...
+		query = query.Joins("JOIN employee_stacks ON employees.id = employee_stacks.employee_id AND employee_stacks.stack_id = ?", filter.StackID)
+	}
+
+	if filter.ProjectID != "" {
+		query = query.Joins("JOIN project_members ON employees.id = project_members.employee_id AND project_members.project_id = ?", filter.ProjectID)
+	}
+
+	if filter.Keyword != "" {
+		query = query.Where("github_id like ? OR discord_id like ? OR notion_id like ? OR full_name like ? OR team_email like ?",
+			fmt.Sprintf("%%%s%%", filter.Keyword), fmt.Sprintf("%%%s%%", filter.Keyword), fmt.Sprintf("%%%s%%", filter.Keyword), fmt.Sprintf("%%%s%%", filter.Keyword), fmt.Sprintf("%%%s%%", filter.Keyword))
```
Comment (nguyennh4522) [inline pkg/store/employee/employee.go:70]: look confused

Comment (datphamcode295) [reply pkg/store/employee/employee.go]: I updated it to Or()

**Review by nguyennh4522 — APPROVED** [review]

**Pull request merged**